dashboard/app/models/census/census_summary.rb

```diff
@@ -78,115 +72,35 @@ def self.submission_has_response(submission, is_high_school)
-  #
-  # In a machine learning setting (such as spam filtering) the likelihoods
-  # would be learned from labeled data (old messages marked a spam or not.)
-  # In our case, we've hand crafted the likelihoods based on what we know
-  # about the data sets.
-  #
-  # For each year, we set our prior to the previous year's posterior after
-  # lowering it a bit. The assumption here is that there is some chance
-  # that a school will stop teaching computer science.
-  #
-
-  #
-  # NOTE: within each deepest likelihood grouping, the weights of all the
-  # options must add up to 1.0
-  #
-  SUBMISSION_LIKELIHOODS = {
-    given_teaches: {
-      survey_yes: 0.8,
-      survey_no: 0.2, # There is a chance people will not be aware of the offering at their school
-    },
-    given_does_not_teach: {
-      survey_yes: 0.1, # Here we are accounting for people misunderstanding / misreporting
-      survey_no: 0.9,
-    },
-  }.freeze
-
-  AP_LIKELIHOODS = {
-    given_teaches: {
-      data_present: 0.2, # Most schools don't have AP even if they teach CS
-      no_data: 0.8,
-    },
-    given_does_not_teach: {
-      data_present: 0.00001, # There is a chance that the school code mappings are wrong
-      no_data: 0.99999,
-    },
-  }.freeze
-
-  IB_LIKELIHOODS = {
-    given_teaches: {
-      data_present: 0.01, # Most schools don't have IB even if they teach CS
-      no_data: 0.99,
-    },
-    given_does_not_teach: {
-      data_present: 0.00001, # There is a chance that the school code mappings are wrong
-      no_data: 0.99999,
-    },
-  }.freeze
-
-  STATE_LIKELIHOODS = {
-    given_teaches: {
-      data_present: 0.99,
-      no_data: 0.01,
-    },
-    given_does_not_teach: {
-      data_present: 0.001,
-      no_data: 0.999,
-    },
-  }.freeze
+  HISTORICAL_RESULTS_MAP = {
+    "YES" => "HISTORICAL_YES",
+    "NO" => "HISTORICAL_NO",
+    "MAYBE" => "HISTORICAL_MAYBE",
+  }
 
-  def self.decide_teaches(posterior)
-    if posterior.nil?
-      nil
-    elsif posterior <= 0.1
-      "N"
-    elsif posterior >= 0.8
-      "Y"
-    else
-      "M"
-    end
-  end
-
-  def self.summarize_school_data(school, school_years, years_with_ap_data, years_with_ib_data, state_years_with_data)
-    summarize_school_data_simple(
-      {
-        school: school,
-        school_years: school_years,
-        years_with_ap_data: years_with_ap_data,
-        years_with_ib_data: years_with_ib_data,
-        state_years_with_data: state_years_with_data,
-      }
-    )
+  def self.map_historical_teaches_cs(historical_value)
+    HISTORICAL_RESULTS_MAP[historical_value]
   end
 
-  def self.summarize_school_data_simple(summarization_data)
+  def self.summarize_school_data(summarization_data)
     school = summarization_data[:school]
     school_years = summarization_data[:school_years]
     state_years_with_data = summarization_data[:state_years_with_data]
 
     summaries = []
+    last_years_result = nil
+    two_years_ago_result = nil
```

Comment:
Where are these populated? They appear to be read a few times starting on line [248](https://github.com/code-dot-org/code-dot-org/pull/21179/files#diff-1d1fe297025f24dc816f52703bdf6e65R248) but only assigned a value (aside from this `nil`) later on line [264](https://github.com/code-dot-org/code-dot-org/pull/21179/files#diff-1d1fe297025f24dc816f52703bdf6e65R264).

Reply:
Yes, that is where they are populated. They are initially `nil` and then updated at the end of each iteration of the loop over school years.

Reply:
OIC. There is so much churn in this PR it was difficult to see this is in a [loop](https://github.com/code-dot-org/code-dot-org/blob/19ee09faee4b97976535c7acd4749ddb37e6b1f3/dashboard/app/models/census/census_summary.rb#L94).

The body of that loop is enormous. Is there a way we can encapsulate that into a function so it's easier to grok. Then the loop could look more like

```ruby
school_years.each do |school_year|
  summary = summarize(school_year, last_years_result, two_years_ago_result)
  summaries.push summary

  two_years_ago_result = last_years_result
  last_years_result = summary.teaches_cs
end
```

With the bulk of the work in `summarize`.
Are we likely to use more than the last 2 years' results? Might this be better as an array: `previous_years_results = []`?

Reply:
Yeah, the diff makes it really hard to see what's happening. I like the idea of refactoring this to make the loop much shorter. Not sure about an array vs. explicitly named variables. I'll try it out and see which approach works better.

Reply:
I've now broken up the method and I did end up using the array approach for the previous years' data.